Automattic/jetpack
# Contact Form: Add Gutenberg based form builder blocks

**Status: In Progress**

Adding a Gutenberg form builder to Jetpack for use in the contact form module.

There are a couple of issues blocking any kind of usable solution being completed, outlined further below.

#### Changes proposed in this Pull Request:

* Add form block based on the InnerBlock component
* Add text, textarea, and button form builder element blocks.

#### Testing instructions:

* Add a Form block to a Gutenberg page or post.
* Default Form should contain 
  * Name, Email textboxes
  * Message textarea, 
  * Submit Button
* Ability to add form elements should only be enabled against the form block (fix for this is blocked) 
* Text elements should have a default label, no placeholder, and be required.
* Buttons should have a default label

#### Proposed changelog entry for your changes:

* Added Gutenberg form builder blocks for the contact form module.

#### Further info:

Blocks:

* Form - The main block, consists of an InnerBlock for the `<form>`
element and a set template consisting of a default set of inner blocks
preconfigured as a contact form
* Text - General `<input type=text>/<textarea>` field with configurable
  label, rows
* Button - A `<button type=submit>` with configurable button text

Blocking Release:

* InnerBlock templating requires WordPress/gutenberg#5452.
* Some form of block inserter restrictions are needed, likely solution coming via WordPress/gutenberg#6346 / WordPress/gutenberg#5540

Todo - Integration with grunion-contact-form:

* Captcha support should be relatively easy to achieve with a serverside
  rendered block
* Serverside controlled form action/method's - probably best implemented [withAPIData](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data)

Todo - Handle duplicate field names gracefully:

* Currently uses a variation on the label / field name to set
  id / for settings on label / inputs.
* Currently uses same variation for `<input name="">` attributes

How it looks:
![](https://i.imgur.com/Y5fDbud.gif)
